presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java

```diff
@@ -38,43 +51,65 @@
 import static com.facebook.presto.server.QueryStateInfo.createQueuedQueryStateInfo;
 import static com.facebook.presto.server.security.RoleType.ADMIN;
 import static com.facebook.presto.server.security.RoleType.USER;
+import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.base.Strings.isNullOrEmpty;
 import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.net.HttpHeaders.X_FORWARDED_PROTO;
 import static java.util.Objects.requireNonNull;
 import static javax.ws.rs.core.Response.Status.NOT_FOUND;
+import static javax.ws.rs.core.Response.Status.SERVICE_UNAVAILABLE;
 
 @Path("/v1/queryState")
 @RolesAllowed({ADMIN, USER})
 public class QueryStateInfoResource
 {
     private final DispatchManager dispatchManager;
     private final ResourceGroupManager<?> resourceGroupManager;
+    private final boolean resourceManagerEnabled;
+    private final InternalNodeManager internalNodeManager;
+    private final Optional<ResourceManagerProxy> proxyHelper;
 
     @Inject
     public QueryStateInfoResource(
             DispatchManager dispatchManager,
-            ResourceGroupManager<?> resourceGroupManager)
+            ResourceGroupManager<?> resourceGroupManager,
+            InternalNodeManager internalNodeManager,
+            ServerConfig serverConfig,
+            Optional<ResourceManagerProxy> proxyHelper)
     {
         this.dispatchManager = requireNonNull(dispatchManager, "dispatchManager is null");
         this.resourceGroupManager = requireNonNull(resourceGroupManager, "resourceGroupManager is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.resourceManagerEnabled = requireNonNull(serverConfig, "serverConfig is null").isResourceManagerEnabled();
+        this.proxyHelper = requireNonNull(proxyHelper, "proxyHelper is null");
     }
 
     @GET
     @Produces(MediaType.APPLICATION_JSON)
-    public List<QueryStateInfo> getQueryStateInfos(@QueryParam("user") String user)
+    public void getQueryStateInfos(@QueryParam("user") String user,
+            @QueryParam("includeLocalQueryOnly") @DefaultValue("false") boolean isIncludeLocalQueryOnly,
+            @HeaderParam(X_FORWARDED_PROTO) String xForwardedProto,
+            @Context UriInfo uriInfo,
+            @Context HttpServletRequest servletRequest,
+            @Suspended AsyncResponse asyncResponse)
     {
-        List<BasicQueryInfo> queryInfos = dispatchManager.getQueries();
-
-        if (!isNullOrEmpty(user)) {
-            queryInfos = queryInfos.stream()
-                    .filter(queryInfo -> Pattern.matches(user, queryInfo.getSession().getUser()))
-                    .collect(toImmutableList());
+        if (resourceManagerEnabled && !isIncludeLocalQueryOnly) {
+            proxyQueryStateInfo(servletRequest, asyncResponse, xForwardedProto, uriInfo);
```

Comment:
instead of pulling all info from RM, we could just use the other coordinator query info from RM and merge it with the existing query info. Not a must but a good to have optimization.

Reply:
I think we can do this optimization. Only caveat is that we need to do back and forth GET-POST conversion while proxying request between coordinator and RM. I will look into the changes required for this.